File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.js
let fs = require('fs')
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n")
//let input = fs.readFileSync("./EX.txt").toString().trim().split("\n")

const N = parseInt(input[0])

const arr = input[1].split(" ").map(x => parseInt(x))
const op = input[2].split(" ").map(x => parseInt(x))

let max = -Infinity
let min = Infinity
function back(n, sm) {
    
    if (n >= N) {
        max = Math.max(sm, max)
        min = Math.min(sm, min)
        return
    }

    for (let i = 0; i < op.length; i++) {
        if (op[i]) {
            op[i] -= 1

            if (i === 0) {
                
                back(n+1, sm+arr[n])
              
            } else if (i === 1) {
       
                back(n+1, sm-arr[n])
       
            } else if (i === 2) {
          
                back(n+1, sm*arr[n])
           
            } else if (i === 3) {
          
                if (sm < 0) {
                    back(n+1, -parseInt(-sm / arr[n]))
                } else {
                    back(n+1, parseInt(sm / arr[n]))
                }
              
            }
            op[i] += 1
        }
    }
}

back(1, arr[0])
console.log(max === -0 ? 0 : max)
console.log(min === -0 ? 0 : min)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
import sys
input = sys.stdin.readline
from itertools import combinations

def back(n, sm):
    global max_val, min_val

    if n >= N:
        max_val = max(max_val, sm)
        min_val = min(min_val, sm)
        return

    for i in range(len(op)):
        if op[i]:
            if i == 0:
                op[i] -= 1
                back(n + 1, sm + arr[n])
                op[i] += 1
            elif i == 1:
                op[i] -= 1
                back(n + 1, sm - arr[n])
                op[i] += 1
            elif i == 2:
                op[i] -= 1
                back(n + 1, sm * arr[n])
                op[i] += 1
            elif i == 3:
                op[i] -= 1
                if sm < 0:
                    back(n + 1, -(-sm // arr[n]))
                else:
                    back(n + 1, sm // arr[n])
                op[i] += 1


N = int(input())
arr = [int(x) for x in input().split()]
op = [int(x) for x in input().split()]

max_val = -float('inf')
min_val = float('inf')
back(1, arr[0])

print(max_val, min_val)